lib/mongodb.js

```diff
@@ -483,6 +482,121 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
 };
 
 /**
+ * Parses the data input for replace operations and returns the
+ * sanitised version of the object.
+ *
+ * @param data
+ * @returns {*}
+ */
+MongoDB.prototype.parseReplaceData = function(model, data, options) {
+  options = options || {};
+  var parsedData = {};
+  
+  var modelClass = this._models[model];
+  
+  var allowExtendedOperators = this.settings.allowExtendedOperators;
+  if (options.hasOwnProperty('allowExtendedOperators')) {
+      allowExtendedOperators = options.allowExtendedOperators === true;
+  } else if (allowExtendedOperators !== false && modelClass.settings.mongodb
+      && modelClass.settings.mongodb.hasOwnProperty('allowExtendedOperators')) {
+      allowExtendedOperators = modelClass.settings.mongodb.allowExtendedOperators === true;
+  } else if (allowExtendedOperators === true) {
+      allowExtendedOperators = true;
+  }
+  
+  if (allowExtendedOperators) {
+    // Check for other operators and sanitize the data obj
+    var acceptedOperators = [
+      // Field operators
+      '$currentDate', '$inc', '$max', '$min', '$mul', '$rename', '$setOnInsert', '$set', '$unset',
+      // Array operators
+      '$addToSet', '$pop', '$pullAll', '$pull', '$pushAll', '$push',
+      // Bitwise operator
+      '$bit'
+    ];
+
+    var usedOperators = 0;
+
+    // each accepted operator will take its place on parsedData if defined
+    for (var i = 0; i < acceptedOperators.length; i++) {
+      if(data[acceptedOperators[i]]) {
+        parsedData[acceptedOperators[i]] = data[acceptedOperators[i]];
+        usedOperators++;
+      }
+    }
+
+    // if parsedData is still empty, then we fallback to $set operator
+    if(usedOperators === 0) {
+      parsedData.$set = data;
+    }
+  } else {
+    parsedData.$set = data;
+  }
+  
+  parsedData.$unset = {};
+  var props = this.getModelDefinition(model).properties;
+  var keys = Object.keys(props);
+  for (var i = 0, n = keys.length; i < n; i++) {
+    var key = keys[i];
+    var p = parsedData.$set[key];
+    if (key != 'id' && p == null) {
+      parsedData.$unset[key] = "";
+    }
+  }
+  return parsedData;
```

Comment:
I think the solution based on `{ $set, $unset }` will not work correctly when the data contains any extra properties not defined in the model. I believe we need to use [Replace the document](https://docs.mongodb.org/manual/tutorial/modify-documents/#replace-the-document) variant instead.

> To replace the entire content of a document except for the _id field, pass an entirely new document as the second argument to update().